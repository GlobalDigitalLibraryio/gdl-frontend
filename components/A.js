// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */
import React from 'react';
import { css, cx } from 'react-emotion';
import theme from '../style/theme';

type Props = {
  className?: string,
  isBold?: boolean,
  isUppercased?: boolean,
  onClick?: (
    event: SyntheticEvent<HTMLButtonElement> | SyntheticEvent<HTMLAnchorElement>
  ) => any
};

const style = css`
  font-size: inherit;
  border: none;
  background: transparent;
  &:hover {
    color: ${theme.colors.blues.dark};
  }
  &[disabled] {
    cursor: not-allowed;
    color: ${theme.colors.gray};
  }
`;

const uppcasedStyle = css`
  text-transform: uppercase;
`;

const boldStyle = css`
  font-weight: 500;
`;

/* eslint-disable jsx-a11y/anchor-has-content */
export default function({ isBold, isUppercased, className, ...props }: Props) {
  return (
    <a
      className={cx(
        style,
        { [uppcasedStyle]: isUppercased, [boldStyle]: isBold },
        className
      )}
      {...props}
    />
  );
}
