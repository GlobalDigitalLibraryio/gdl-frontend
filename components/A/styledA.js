// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */
import styled, { css } from 'react-emotion';
import theme from 'styled-theming';

import rotate360 from '../../style/rotate360';
import { colors, fonts } from '../../style/theme';

const color = theme('category', {
  library: css`
    color: ${colors.link.default};
    &:hover {
      color: ${colors.link.defaultHover};
    }
  `,
  classroom: css`
    color: ${colors.link.alternate};
    &:hover {
      color: ${colors.link.alternateHover};
    }
  `
});

export const A = styled('a')`
  ${color};
  font-size: inherit;
  border: none;
  background: transparent;
  &[disabled] {
    cursor: not-allowed;
    color: ${colors.base.gray};
  }
`;

export const uppcasedStyle = css`
  text-transform: uppercase;
`;

export const boldStyle = css`
  font-weight: ${fonts.weight.medium};
`;

export const underlineStyle = css`
  text-decoration: underline;
`;

export const loadingStyle = css`
  color: transparent;
  text-shadow: none;
  position: relative;
  pointer-events: none;
  &:after {
    animation: ${rotate360} 500ms infinite linear;
    border: 2px solid ${colors.base.black};
    border-radius: 100px;
    border-right-color: transparent;
    border-top-color: transparent;
    content: '';
    display: block;
    width: 0.8em;
    height: 0.8em;
    position: absolute;
    left: calc(50% - (0.8em / 2));
    top: calc(50% - (0.8em / 2));
  }
`;
