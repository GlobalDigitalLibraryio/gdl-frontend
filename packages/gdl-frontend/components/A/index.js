// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */
import React, { type Node } from 'react';
import { cx } from 'react-emotion';
import {
  A,
  uppcasedStyle,
  boldStyle,
  underlineStyle,
  loadingStyle
} from './styledA';

type Props = {|
  'aria-expanded'?: boolean,
  children: Node,
  className?: string,
  disabled?: boolean,
  isBold?: boolean,
  isUppercased?: boolean,
  isUnderlined?: boolean,
  isLoading?: boolean,
  href?: string,
  target?: '_blank',
  rel?: 'noopener noreferrer',
  style?: {},
  onClick?: (
    event: SyntheticEvent<HTMLButtonElement> | SyntheticEvent<HTMLAnchorElement>
  ) => any
|};

const Button = A.withComponent('button');

/**
 * Renders anchor is href is passed, renders button if no href is passed
 */
export default function({
  isBold,
  isUppercased,
  isUnderlined,
  isLoading,
  className,
  href,
  ...props
}: Props) {
  const Elem = href != null ? A : Button;
  return (
    <Elem
      href={href}
      className={cx(
        {
          [uppcasedStyle]: isUppercased,
          [boldStyle]: isBold,
          [underlineStyle]: isUnderlined,
          [loadingStyle]: isLoading
        },
        className
      )}
      {...props}
    />
  );
}
