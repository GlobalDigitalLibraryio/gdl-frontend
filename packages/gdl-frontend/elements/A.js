// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2018 GDL
 *
 * See LICENSE
 */

import React from 'react';
import { Typography } from '@material-ui/core';
import { css, cx } from 'react-emotion';

type Props = {
  className?: string,
  openNewTab?: boolean
};

const A = ({ openNewTab, className, ...restProps }: Props) => {
  const props = {
    ...restProps
  };
  if (openNewTab) {
    // $FlowFixMe
    props.target = '_blank';
    // $FlowFixMe
    props.rel = 'noopener noreferrer';
  }
  return (
    <Typography
      component="a"
      color="primary"
      className={cx(style, className)}
      {...props}
    />
  );
};

const style = css`
  &:hover {
    text-decoration: underline;
  }
}`;

export default A;
