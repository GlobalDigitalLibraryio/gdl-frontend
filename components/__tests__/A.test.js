// @flow
/**
 * Part of GDL gdl-frontend.
 * Copyright (C) 2017 GDL
 *
 * See LICENSE
 */

import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { createSerializer } from 'jest-emotion';
import * as emotion from 'emotion';
import A from '../A';

expect.addSnapshotSerializer(createSerializer(emotion))

test('Is underlined anchor', () => {
  const tree = shallow(<A>Link</A>);

  expect(toJson(tree)).toMatchSnapshot();
});
