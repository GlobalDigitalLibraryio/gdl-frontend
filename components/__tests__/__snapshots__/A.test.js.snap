// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Is underlined anchor 1`] = `
.emotion-0 {
  text-decoration: underline;
}

<a
  className="emotion-0"
>
  Link
</a>
`;
